test/math_libraries/convolutions.py

```diff
@@ -0,0 +1,100 @@
+import torch
+from torch.testing._internal.common_device_type import instantiate_device_type_tests, \
+    dtypes, dtypesIfCUDA, skipCUDAIfRocm, onlyOnCPUAndCUDA
+from torch.testing._internal.common_nn import NNTestCase
+from torch.testing._internal.common_utils import run_tests
+import os
+import json
+
+class TestConvExt(NNTestCase):
+    def _subtensor(self, tensor, dim, group, g):
+        if tensor is None:
+            return None
+        group_size = int(tensor.size(dim) / group)
+        return tensor.narrow(dim, group_size * g, group_size).contiguous()
+
+    def _thnn_conv(self, input, weight, k, bias, s, p, d):
+        if d[0] > 1 or d[1] > 1:
+            return torch._C._nn.slow_conv_dilated2d(input, weight, k, bias, s, p, d)
+        else:
+            return torch._C._nn.thnn_conv2d(input, weight, k, bias, s, p)
+
+    def _thnn_conv_group(self, input, weight, k, bias, s, p, d, group):
+        if group == 1:
+            return self._thnn_conv(input, weight, k, bias, s, p, d)
+        else:
+            outputs = []
+            for g in range(group):
+                input_g = self._subtensor(input, 1, group, g)
+                weight_g = self._subtensor(weight, 0, group, g)
+                bias_g = self._subtensor(bias, 0, group, g)
+                outputs.append(self._thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+            return torch.cat(outputs, 1)
+
+    # Set multiple situations to improve convolution2d test coverage
+    @onlyOnCPUAndCUDA
+    @skipCUDAIfRocm
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float)
+    def test_conv2d_ext(self, device, dtype):
+        case_files = os.popen('ls shapes_*').read().strip().split()
+        total_cases = list()
+        for file_name in case_files:
+            with open(file_name) as f:
+                model_cases = json.loads(f.read())
+            total_cases += model_cases
+        for case in total_cases:
+            bs = case['mb']
+            group = case['g']
+            ic, ih, iw = case['ic'], case['ih'], case['iw']
+            oc = case['oc']
+            kh, kw = case['kh'], case['kw']
+            sh, sw = case['sh'], case['sw']
+            ph, pw = case['ph'], case['pw']
+            dh, dw = case['dh'], case['dw']
+            has_bias = case['bias']
+            if dh == 0 or dw == 0:
+                continue
+
+            ic_g = ic // group
+            torch.manual_seed(1)
+            input = torch.randn((bs, ic, ih, iw), device=device, dtype=dtype, requires_grad=True)
+            torch.manual_seed(2)
+            weight = torch.randn((oc, ic_g, kh, kw), device=device, dtype=dtype) * 0.01
+            torch.manual_seed(3)
+            bias = None if has_bias else torch.randn((oc), device=device, dtype=dtype)
```

Comment:
This seems incorrect, bias should not be None if has_bias? In json the has_bias value is a string, not bool, so in python comparisons has_bias will always be True regardless of the string in json, and bias will always be None.